controllers/containerimagebuild_resources.go

```diff
@@ -0,0 +1,182 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	forgev1alpha1 "github.com/dominodatalab/forge/api/v1alpha1"
+)
+
+func (r *ContainerImageBuildReconciler) checkPrerequisites(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	if err := r.checkServiceAccount(ctx, cib); err != nil {
+		return err
+	}
+	if err := r.checkRole(ctx, cib); err != nil {
+		return err
+	}
+	if err := r.checkRoleBinding(ctx, cib); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ContainerImageBuildReconciler) checkServiceAccount(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	err := r.Get(ctx, types.NamespacedName{Name: cib.Name, Namespace: cib.Namespace}, &corev1.ServiceAccount{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cib.Name,
+			Namespace: cib.Namespace,
+			Labels:    cib.Labels,
+		},
+	}
+	if err := controllerutil.SetControllerReference(cib, sa, r.Scheme); err != nil {
+		return err
+	}
+
+	return r.Create(ctx, sa)
+}
+
+func (r *ContainerImageBuildReconciler) checkRole(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	err := r.Get(ctx, types.NamespacedName{Name: cib.Name, Namespace: cib.Namespace}, &rbacv1.Role{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cib.Name,
+			Namespace: cib.Namespace,
+			Labels:    cib.Labels,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{forgev1alpha1.GroupVersion.Group},
+				Resources: []string{"containerimagebuilds"},
+				Verbs:     []string{"get"},
+			},
+			{
+				APIGroups: []string{forgev1alpha1.GroupVersion.Group},
+				Resources: []string{"containerimagebuilds/status"},
+				Verbs:     []string{"update"},
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"secrets"},
+				Verbs:     []string{"get"},
+			},
+		},
+	}
+	if err := controllerutil.SetControllerReference(cib, role, r.Scheme); err != nil {
+		return err
+	}
+
+	return r.Create(ctx, role)
+}
+
+func (r *ContainerImageBuildReconciler) checkRoleBinding(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	err := r.Get(ctx, types.NamespacedName{Name: cib.Name, Namespace: cib.Namespace}, &rbacv1.RoleBinding{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
```

Comment:
Definitely would be nice to avoid this boilerplate & the reference setting if it isn't crazy

Reply:
I see, you mean extract this into a reusable func? If so, that makes sense to me.

Reply:
Yep, if possible

Reply:
See https://github.com/dominodatalab/forge/pull/42/commits/0cd4662fd4ac9ff6e00427b1164926059f9d2478